### PR TITLE
Story/vendorize dependencies

### PR DESCRIPTION
**Pull request by mmullin-halcyon**
VENDORIZE: added zlib and elfutils as vendored copies
    When the static_dependencies feature is applied,
    libbpf-sys will now compile static versions of elfutils and zlib,
    and will link against those newly compiled static libraries.

    Justification: Many mainstream distributions are not shipping static versions
    of libelf or zlib (eg Fedora).  Including them in libbpf-sys
    makes developing rust bpf applications much easier than forcing
    developers to do all this work themselves.

    signed-off-by: Michael Mullin <mmullin@halcyon.ai>

### CODE REVIEW

File: Cargo.toml
Comment (danielocfb) [inline Cargo.toml:36]: I feel like `vendored` is more commonly used (and more succinct) terminology. E.g., [here](https://github.com/sfackler/rust-openssl/blob/a71f492fccaf0743e987cdf12f7032683d951ae9/openssl-sys/Cargo.toml#L18).

File: build.rs
Comment (danielocfb) [inline build.rs:231]: Nit: move declaration closer to the location of use for the `configure` command?

**Review by danielocfb — APPROVED** [review]
Not the maintainer, but seems like a useful feature to me! Thanks. Left a few comments.

File: build.rs
Comment (danielocfb) [reply build.rs:104]: That's fine then.

File: build.rs
Comment (mmullin-halcyon) [reply build.rs]: done

File: Cargo.toml
Comment (mmullin-halcyon) [reply Cargo.toml]: done

File: build.rs
Comment (mmullin-halcyon) [reply build.rs]: done

File: build.rs
Comment (alexforster) [inline build.rs:182]: Have you actually seen this happen? I think Cargo will prevent this.

File: .gitmodules
Comment (alexforster) [inline .gitmodules:6]: nit: I'd prefer to use the HTTP transport, like `https://sourceware.org/git/elfutils.git`

https://sourceware.org/git/?p=elfutils.git;a=summary

File: zlib
Comment (alexforster) [inline zlib:1]: Note to future me: commit references `zlib 1.2.13`

File: elfutils
Comment (alexforster) [inline elfutils:1]: Note to future me: commit references `elfutils 0.189`

**Comment by alexforster** [comment]
I can't get this to build. Commands to reproduce:

```
git clone --recurse-submodules https://github.com/mmullin-halcyon/libbpf-sys.git
cd libbpf-sys
git checkout story/vendorize_dependencies
docker run --rm -v "$(pwd):/root" -it debian:latest bash
cd /root
apt-get update
apt-get -y install curl git build-essential pkgconf
curl --proto '=https' --tlsv1.2 -sSf https://sh.rustup.rs | sh -s -- -y
source "$HOME/.cargo/env"
cargo build --features vendored
```

Output:

```
error: failed to run custom build command for `libbpf-sys v1.1.1+v1.1.0 (/root)`

Caused by:
  process didn't exit successfully: `/root/target/debug/build/libbpf-sys-5d736d1453869834/build-script-build` (exit status: 101)
  --- stdout
  OPT_LEVEL = Some("0")
  TARGET = Some("x86_64-unknown-linux-gnu")
  HOST = Some("x86_64-unknown-linux-gnu")
  cargo:rerun-if-env-changed=CC_x86_64-unknown-linux-gnu
  CC_x86_64-unknown-linux-gnu = None
  cargo:rerun-if-env-changed=CC_x86_64_unknown_linux_gnu
  CC_x86_64_unknown_linux_gnu = None
  cargo:rerun-if-env-changed=HOST_CC
  HOST_CC = None
  cargo:rerun-if-env-changed=CC
  CC = None
  cargo:rerun-if-env-changed=CFLAGS_x86_64-unknown-linux-gnu
  CFLAGS_x86_64-unknown-linux-gnu = None
  cargo:rerun-if-env-changed=CFLAGS_x86_64_unknown_linux_gnu
  CFLAGS_x86_64_unknown_linux_gnu = None
  cargo:rerun-if-env-changed=HOST_CFLAGS
  HOST_CFLAGS = None
  cargo:rerun-if-env-changed=CFLAGS
  CFLAGS = None
  cargo:rerun-if-env-changed=CRATE_CC_NO_DEFAULTS
  CRATE_CC_NO_DEFAULTS = None
  DEBUG = Some("true")
  CARGO_CFG_TARGET_FEATURE = Some("fxsr,sse,sse2")

  --- stderr
  make: *** No targets specified and no makefile found.  Stop.
  Please specify at least one package name on the command line.
  thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: Os { code: 2, kind: NotFound, message: "No such file or directory" }', build.rs:183:66
  note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

File: build.rs
Comment (mmullin-halcyon) [reply build.rs:182]: I have not see issues compiling either zlib nor libbpf.

However, during development prior to using a feature flag, elfutils had a 100% reproducible problem when compiling a program which uses libbpf-rs as a dependency and libbpf-cargo as a build-dependency.

File: .gitmodules
Comment (mmullin-halcyon) [reply .gitmodules]: After changing this to url = https://sourceware.org/git/elfutils.git

I experience the following issue
```

$ cargo update
    Updating git repository `https://github.com/mmullin-halcyon/libbpf-sys.git`
    Updating git submodule `https://sourceware.org/git/elfutils.git`
warning: spurious network error (3 tries remaining): early EOF; class=Net (12); code=Eof (-20)
warning: spurious network error (2 tries remaining): early EOF; class=Net (12); code=Eof (-20)
warning: spurious network error (1 tries remaining): early EOF; class=Net (12); code=Eof (-20)
error: failed to get `libbpf-sys` as a dependency of package `ttt v0.1.0 (/code/ttt)`

Caused by:
  failed to load source for dependency `libbpf-sys`

Caused by:
  Unable to update https://github.com/mmullin-halcyon/libbpf-sys.git?branch=story/vendorize_dependencies

Caused by:
  failed to update submodule `elfutils`

Caused by:
  failed to fetch submodule `elfutils` from https://sourceware.org/git/elfutils.git

Caused by:
  network failure seems to have happened
  if a proxy or similar is necessary `net.git-fetch-with-cli` may help here
  https://doc.rust-lang.org/cargo/reference/config.html#netgit-fetch-with-cli

Caused by:
  early EOF; class=Net (12); code=Eof (-20)
```

I have tested this on multiple machines by doing the following

$ cargo new ttt && echo "libbpf-sys = { git = \"https://github.com/mmullin-halcyon/libbpf-sys.git\", branch = \"story/vendorize_dependencies\" }" >> ttt/Cargo.toml && cd ttt && cargo build

using ```CARGO_NET_GIT_FETCH_WITH_CLI=true cargo update``` can fix this problem, but it's not a good experience.

I think it might be a better experience to go back to the git://sourceware.org/git/elfutils.git

**Comment by mmullin-halcyon** [comment]
Addressed compilation issue && poor user experience if compile tools are unavailable

File: .gitmodules
Comment (alexforster) [reply .gitmodules]: Hmm, weird. It can stay as-is, then.

Comment (alexforster) [reply .gitmodules]: Or rather, it looks like you changed it to `https`. Feel free to change it back to `git`.

Comment (mmullin-halcyon) [reply .gitmodules:6]: done

**Comment by alexforster** [comment]
Sorry for taking so long, I think this looks good now. Going to merge this, and after I merge the other PR that bumps the libbpf version I'll cut a release.

**Pull request merged**
